src/check/parse/tokenize/src/main.zig

```diff
@@ -0,0 +1,1399 @@
+                    try self.output.pushToken(T.LowerIdent, start, len);
+                },
+
+                // Uppercase identifiers
+                'A'...'Z' => {
+                    _ = self.cursor.chompIdentUpper();
+                    const len = self.cursor.pos - start;
+                    try self.output.pushToken(T.UpperIdent, start, len);
+                },
+
+                // String-like literal starting with a single or double quote
+                '"', '\'' => {
+                    // Note this may return StringBegin/StringPart instead of String,
+                    // in the case of a string interpolation.
+                    const tok = try self.tokenizeStringLikeLiteral(b);
+                    try self.output.pushToken(tok, start, self.cursor.pos - start);
+                },
+
+                // first byte of a UTF-8 sequence
+                0b11000000...0xff => {
+                    const info = self.cursor.peekUnicode();
+                    switch (info.kind) {
+                        .LetterUpper => {
+                            self.cursor.pos += info.length;
+                            self.cursor.chompIdentGeneral();
+                            const len = self.cursor.pos - start;
+                            try self.output.pushToken(T.UpperIdent, start, len);
+                        },
+                        .LetterNotUpper => {
+                            self.cursor.pos += info.length;
+                            self.cursor.chompIdentGeneral();
+                            const len = self.cursor.pos - start;
+                            try self.output.pushToken(T.LowerIdent, start, len);
+                        },
+                        else => {
+                            self.cursor.pos += info.length;
+                            self.cursor.pushMessageHere(.UnknownToken);
+                        },
+                    }
+                },
+
+                // Fallback for any unknown token.
+                else => {
+                    self.cursor.pushMessageHere(.UnknownToken);
+                    self.cursor.pos += 1;
+                },
+            }
+        }
+
+        if (self.output.lines.items[self.output.lines.items.len - 1].token_start == self.output.tokens.len) {
+            _ = self.output.lines.pop();
+        }
+
+        try self.pushToken(T.EndOfFile, 0);
+    }
+
+    pub fn tokenizeStringLikeLiteral(self: *Tokenizer, term: u8) !T {
+        const start = self.cursor.pos;
+        // Skip the initial quote.
+        self.cursor.pos += 1;
+        var multiline: bool = false;
+        if (self.cursor.peek() == term and self.cursor.peekAt(1) == term) {
+            self.cursor.pos += 2;
+            multiline = true;
+        }
+        return try self.tokenizeStringLikeLiteralBody(false, term, start, multiline);
+    }
+
+    pub fn tokenizeStringLikeLiteralBody(self: *Tokenizer, already_started: bool, term: u8, start: usize, multiline: bool) !T {
```

Comment:
I think it would be really nice to change the bool inputs to this functions into enums. It is really hard to read a call to this function and understand the args (espically with two bool inputs)